source/logic.rst

```diff
@@ -0,0 +1,1223 @@
+See if you can understand the strategy we have outlined
+and finish the proof.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def converges_to (s : ℕ → ℝ) (a : ℝ) :=
+    ∀ ε, 0 < ε → ∃ N, ∀ n, N ≤ n → abs (s n - a) < ε
+
+    variables {s t : ℕ → ℝ} {a : ℝ}
+
+    theorem exists_abs_le_of_converges_to (cs : converges_to s a) :
+      ∃ N b, ∀ n, N ≤ n → abs (s n) < b :=
+    sorry
+
+    -- BEGIN
+    lemma aux (cs : converges_to s a) (ct : converges_to t 0) :
+      converges_to (λ n, s n * t n) 0 :=
+    begin
+      intros ε εpos, dsimp,
+      rcases exists_abs_le_of_converges_to cs with ⟨N₀, B, h₀⟩,
+      have Bpos : 0 < B,
+        from lt_of_le_of_lt (abs_nonneg _) (h₀ N₀ (le_refl _)),
+      have pos₀ : ε / B > 0,
+        from div_pos εpos Bpos,
+      cases ct _ pos₀ with N₁ h₁,
+      sorry
+    end
+    -- END
+
+If you have made it this far, congratulations!
+We are now within striking distance of our theorem.
+The following proof finishes it off.
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    def converges_to (s : ℕ → ℝ) (a : ℝ) :=
+    ∀ ε, 0 < ε → ∃ N, ∀ n, N ≤ n → abs (s n - a) < ε
+
+    theorem converges_to_const (a : ℝ) : converges_to (λ x : ℕ, a) a :=
+    sorry
+
+    variables {s t : ℕ → ℝ} {a b : ℝ}
+
+    theorem converges_to_add
+      (cs : converges_to s a) (ct : converges_to t b):
+    converges_to (λ n, s n + t n) (a + b) :=
+    sorry
+
+    theorem converges_to_mul_const
+      (c : ℝ) (cs : converges_to s a) :
+    converges_to (λ n, c * s n) (c * a) :=
+    sorry
+
+    lemma aux (cs : converges_to s a) (ct : converges_to t 0) :
+      converges_to (λ n, s n * t n) 0 :=
+    sorry
+
+    -- BEGIN
+    theorem converges_to_mul
+      (cs : converges_to s a) (ct : converges_to t b):
+    converges_to (λ n, s n * t n) (a * b) :=
+    begin
+      have h₁ : converges_to (λ n, s n * (t n - b)) 0,
+      { apply aux cs,
+        convert converges_to_add ct (converges_to_const (-b)),
+        ring },
+      convert (converges_to_add h₁ (converges_to_mul_const b cs)),
+      { ext, ring },
+      ring
+    end
+    -- END
+
+We close the section with the observation that our proofs can be generalized.
+For example, the only properties that we have used of the
+natural numbers is that their structure carries a partial order
+with ``min`` and ``max``.
+You can check that everything still works if you replace ``ℕ``
+everywhere by any linear order ``α``:
+
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    variables {α : Type*} [linear_order α]
+
+    def converges_to (s : α → ℝ) (a : ℝ) :=
+    ∀ ε, 0 < ε → ∃ N, ∀ n, N ≤ n → abs (s n - a) < ε
+    -- END
+
+.. TODO: reference to later chapter
+
+In a later chapter, we will see that mathlib has mechanisms
+for dealing with convergence in vastly more general terms,
+not only abstracting away particular features of the domain
+and codomain,
+but also abstracting over different *modes* of convergence.
```

Comment:
Is this "mode" standard terminology?

Reply:
Not really. Maybe "different kinds of convergence"?